File: packages/core/python/caracalai_core/logging.py
"""
Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
Caracal, a product of Garudex Labs

Centralized structured logging redacts sensitive fields for Caracal Python services and SDKs.
"""

from __future__ import annotations

import atexit
import contextvars
import json
import logging
import logging.handlers
import os
import queue
import re
import signal
import socket
import sys
import time
import traceback
from typing import Any, Callable, Mapping

SECRET_KEYS: tuple[str, ...] = (
    "password",
    "passwd",
    "secret",
    "token",
    "api_key",
    "apikey",
    "access_token",
    "refresh_token",
    "id_token",
    "authorization",
    "auth",
    "cookie",
    "set-cookie",
    "client_secret",
    "private_key",
    "signing_key",
    "audit_hmac_key",
    "hmac_key",
    "encryption_key",
    "session",
    "credential",
    "credentials",
)

REDACT_VALUE = "***"

_LEVELS = {"debug": 10, "info": 20, "warn": 30, "warning": 30, "error": 40, "fatal": 50, "critical": 50}

_BEARER_RE = re.compile(r"bearer\s+[A-Za-z0-9._\-+/=]{8,}", re.IGNORECASE)
_JWT_RE = re.compile(r"eyJ[A-Za-z0-9_\-]{8,}\.[A-Za-z0-9_\-]{8,}\.[A-Za-z0-9_\-]{8,}")
_AWS_AKIA_RE = re.compile(r"AKIA[0-9A-Z]{16}")
_AWS_ASIA_RE = re.compile(r"ASIA[0-9A-Z]{16}")
_GCP_KEY_RE = re.compile(r"AIza[0-9A-Za-z_\-]{35}")
_GITHUB_PAT_RE = re.compile(r"gh[pousr]_[A-Za-z0-9]{36,255}")
_SLACK_TOKEN_RE = re.compile(r"xox[abprs]-[A-Za-z0-9\-]{10,}")
_PEM_BLOCK_RE = re.compile(
    r"-----BEGIN [A-Z ]+PRIVATE KEY-----.*?-----END [A-Z ]+PRIVATE KEY-----",
    re.DOTALL,
)

MAX_FIELD_BYTES = int(os.environ.get("CARACAL_LOG_MAX_FIELD_BYTES", "8192"))
_DEBUG_SAMPLE_N = max(1, int(os.environ.get("CARACAL_LOG_SAMPLE_DEBUG", "1")))


def is_secret_key(key: str) -> bool:
    if not key:
        return False
    lk = key.lower()
    return any(s in lk for s in SECRET_KEYS)


def redact_string(s: str) -> str:
    """Scrub bearer tokens, JWTs, and common cloud-secret patterns."""
    if len(s) < 16:
        return s
    s = _PEM_BLOCK_RE.sub(REDACT_VALUE, s)
    s = _BEARER_RE.sub(f"Bearer {REDACT_VALUE}", s)
    s = _JWT_RE.sub(REDACT_VALUE, s)
    s = _AWS_AKIA_RE.sub(REDACT_VALUE, s)
    s = _AWS_ASIA_RE.sub(REDACT_VALUE, s)
    s = _GCP_KEY_RE.sub(REDACT_VALUE, s)
    s = _GITHUB_PAT_RE.sub(REDACT_VALUE, s)
    s = _SLACK_TOKEN_RE.sub(REDACT_VALUE, s)
    return s


def truncate_string(s: str) -> str:
    if MAX_FIELD_BYTES <= 0 or len(s) <= MAX_FIELD_BYTES:
        return s
    return s[:MAX_FIELD_BYTES] + "…[truncated]"


def _serialize_exception(exc: BaseException) -> dict[str, Any]:
    out: dict[str, Any] = {
        "name": type(exc).__name__,
        "message": str(exc),
        "stack": "".join(traceback.format_exception(type(exc), exc, exc.__traceback__)),
    }
    if exc.__cause__ is not None:
        out["cause"] = _serialize_exception(exc.__cause__)
    return out


def redact(value: Any) -> Any:
    if isinstance(value, BaseException):
        return _serialize_exception(value)
    if isinstance(value, str):
        return truncate_string(redact_string(value))
    if isinstance(value, Mapping):
        return {k: (REDACT_VALUE if is_secret_key(str(k)) else redact(v)) for k, v in value.items()}
    if isinstance(value, (list, tuple)):
        seq = [redact(v) for v in value]
        return seq if isinstance(value, list) else tuple(seq)
    return value


_trace_ctx: contextvars.ContextVar[dict[str, str] | None] = contextvars.ContextVar(
    "caracal_trace", default=None
)


def bind_trace(trace_id: str | None = None, span_id: str | None = None) -> contextvars.Token:
    """Bind trace identifiers to the current async/sync context; returns a token
    that can be passed to reset_trace to restore the previous value."""
    payload: dict[str, str] = {}
    if trace_id:
        payload["trace_id"] = trace_id
    if span_id:
        payload["span_id"] = span_id
    return _trace_ctx.set(payload or None)


def reset_trace(token: contextvars.Token) -> None:
    _trace_ctx.reset(token)


def current_trace() -> dict[str, str]:
    cur = _trace_ctx.get()
    return dict(cur) if cur else {}


def parse_traceparent(header: str | None) -> dict[str, str]:
    """Decode a W3C traceparent header into trace_id/span_id keys; empty on parse failure."""
    if not header:
        return {}
    parts = header.split("-")
    if len(parts) < 4 or len(parts[1]) != 32 or len(parts[2]) != 16:
        return {}
    return {"trace_id": parts[1], "span_id": parts[2]}


def _process_base_fields(service: str) -> dict[str, Any]:
    try:
        host = socket.gethostname() or "unknown"
    except OSError:
        host = "unknown"
    return {
        "service": service,
        "hostname": host,
        "pid": os.getpid(),
        "version": os.environ.get("CARACAL_VERSION", "dev"),
        "env": os.environ.get("CARACAL_ENV") or os.environ.get("APP_ENV") or "development",
    }


class _JsonFormatter(logging.Formatter):
    def format(self, record: logging.LogRecord) -> str:
        payload: dict[str, Any] = {
            "level": record.levelname.lower(),
            "time": time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime(record.created))
            + f".{int(record.msecs):03d}Z",
        }
        base = getattr(record, "_caracal_base", None)
        if isinstance(base, Mapping):
            payload.update(base)
        payload["msg"] = record.getMessage()
        bound = getattr(record, "_caracal_bound", None)
        if isinstance(bound, Mapping):
            for k, v in redact(dict(bound)).items():
                payload[k] = v
        extra = getattr(record, "_caracal_extra", None)
        if isinstance(extra, Mapping):
            for k, v in redact(dict(extra)).items():
                payload[k] = v
        if record.exc_info and record.exc_info[1] is not None:
            payload["error"] = _serialize_exception(record.exc_info[1])
        return json.dumps(payload, default=str, separators=(",", ":"))


# Process-wide queue infrastructure: log records produced by application threads
# are pushed onto a bounded queue by QueueHandler (cheap, no formatting), and a
# background QueueListener thread formats and writes to stderr.
_QUEUE_MAXSIZE = int(os.environ.get("CARACAL_LOG_QUEUE_SIZE", "16384"))
_log_queue: "queue.Queue[logging.LogRecord]" = queue.Queue(maxsize=_QUEUE_MAXSIZE)
_listener: logging.handlers.QueueListener | None = None
_dropped = 0
_emitted = 0
_sampled = 0
_debug_counter = 0


def _ensure_listener() -> None:
    global _listener
    if _listener is not None:
        return
    handler = _DynamicStderrHandler()
    handler.setFormatter(_JsonFormatter())
    _listener = logging.handlers.QueueListener(_log_queue, handler, respect_handler_level=True)
    _listener.start()
    atexit.register(_shutdown_listener)


class _DynamicStderrHandler(logging.Handler):
    """StreamHandler that resolves sys.stderr at emit time so test redirection works."""

    def emit(self, record: logging.LogRecord) -> None:
        try:
            msg = self.format(record)
            stream = sys.stderr
            stream.write(msg + "\n")
            stream.flush()
        except Exception:  # noqa: BLE001
            self.handleError(record)


def _shutdown_listener() -> None:
    global _listener
    listener = _listener
    _listener = None
    if listener is not None:
        listener.stop()


def flush_for_test() -> None:
    """Block until the background queue is drained; intended for tests only."""
    _log_queue.join() if hasattr(_log_queue, "join") and False else None
    # QueueListener's internal queue does not support join(); poll instead.
    deadline = time.time() + 1.0
    while time.time() < deadline:
        if _log_queue.empty():
            return
        time.sleep(0.005)


class _NonBlockingQueueHandler(logging.Handler):
    """Pushes records onto the shared queue without blocking; drops when full."""

    def emit(self, record: logging.LogRecord) -> None:
        global _dropped, _emitted
        try:
            _log_queue.put_nowait(record)
            _emitted += 1
        except queue.Full:
            _dropped += 1


class DevLogger:
    """Structured JSON dev/diagnostics logger.

    Two streams in Caracal: this one (developer/infra) and AuditClient
    (immutable end-user audit). Records carry per-service base fields and are
    handed off to a background QueueListener so caller threads never block on
    stderr writes.
    """

    def __init__(self, service: str, level: str | int = "info", bound: Mapping[str, Any] | None = None) -> None:
        self._service = service
        self._bound: dict[str, Any] = dict(bound or {})
        self._base = _process_base_fields(service)
        _ensure_listener()
        self._logger = logging.getLogger(f"caracal.{service}")
        if not any(isinstance(h, _NonBlockingQueueHandler) for h in self._logger.handlers):
            self._logger.handlers.clear()
            self._logger.addHandler(_NonBlockingQueueHandler())
            self._logger.propagate = False
        self.set_level(level)

    def set_level(self, level: str | int) -> None:
        if isinstance(level, str):
            level = _LEVELS.get(level.lower(), logging.INFO)
        self._logger.setLevel(level)

    def with_(self, **fields: Any) -> "DevLogger":
        merged = dict(self._bound)
        merged.update(fields)
        child = DevLogger.__new__(DevLogger)
        child._service = self._service
        child._bound = merged
        child._base = self._base
        child._logger = self._logger
        return child

    def _emit(self, level: int, msg: str, fields: Mapping[str, Any] | None) -> None:
        global _debug_counter, _sampled
        if not self._logger.isEnabledFor(level):
            return
        if level == logging.DEBUG and _DEBUG_SAMPLE_N > 1:
            _debug_counter += 1
            if _debug_counter % _DEBUG_SAMPLE_N != 0:
                _sampled += 1
                return
        exc_info = None
        if fields:
            err = fields.get("err") or fields.get("error") or fields.get("exception")
            if isinstance(err, BaseException):
                exc_info = (type(err), err, err.__traceback__)
        record = self._logger.makeRecord(self._logger.name, level, "", 0, msg, (), exc_info)
        record._caracal_base = self._base  # type: ignore[attr-defined]
        merged_bound = dict(self._bound)
        merged_bound.update(current_trace())
        record._caracal_bound = merged_bound  # type: ignore[attr-defined]
        record._caracal_extra = dict(fields) if fields else None  # type: ignore[attr-defined]
        self._logger.handle(record)

    def debug(self, msg: str, **fields: Any) -> None: self._emit(logging.DEBUG, msg, fields)
    def info(self, msg: str, **fields: Any) -> None: self._emit(logging.INFO, msg, fields)
    def warn(self, msg: str, **fields: Any) -> None: self._emit(logging.WARNING, msg, fields)
    def warning(self, msg: str, **fields: Any) -> None: self._emit(logging.WARNING, msg, fields)
    def error(self, msg: str, **fields: Any) -> None: self._emit(logging.ERROR, msg, fields)
    def fatal(self, msg: str, **fields: Any) -> None: self._emit(logging.CRITICAL, msg, fields)


def shutdown_logging() -> None:
    """Flush and stop the background log listener; safe to call multiple times."""
    _shutdown_listener()


def dropped_log_records() -> int:
    """Number of log records dropped because the background queue was full."""
    return _dropped


def dev_log_metrics() -> dict[str, int]:
    """Snapshot of dev-log counters for /metrics exposure."""
    return {
        "emitted": _emitted,
        "dropped": _dropped,
        "sampled": _sampled,
        "queue_depth": _log_queue.qsize(),
        "queue_cap": _QUEUE_MAXSIZE,
    }


def install_shutdown_handler(extra: Callable[[], None] | None = None, timeout: float = 2.0) -> None:
    """Wire SIGTERM/SIGINT to flush dev logs (and optionally invoke `extra`,
    typically AuditClient.close) before the process exits."""

    def _handler(signum: int, _frame: Any) -> None:
        try:
            if extra is not None:
                extra()
        finally:
            shutdown_logging()
            raise SystemExit(128 + signum)

    signal.signal(signal.SIGTERM, _handler)
    signal.signal(signal.SIGINT, _handler)


def create_logger(service: str, level: str | int | None = None) -> DevLogger:
    if level is None:
        level = os.environ.get("CARACAL_LOG_LEVEL") or os.environ.get("LOG_LEVEL") or "info"
    return DevLogger(service, level)

File: packages/core/ts/src/logging.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Structured JSON logger for TypeScript services.

import { AsyncLocalStorage } from 'node:async_hooks';
import { hostname } from 'node:os';
import type { Writable } from 'node:stream';

type Level = 'debug' | 'info' | 'warn' | 'error' | 'fatal';

const LEVEL_VALUE: Record<Level, number> = { debug: 10, info: 20, warn: 30, error: 40, fatal: 50 };

/**
 * Field names whose values must never appear in dev logs. Mirror this list in
 * packages/core/go/logging/redact.go and packages/core/python/caracalai_core/logging.py.
 */
export const SECRET_KEYS: readonly string[] = Object.freeze([
  'password',
  'secret',
  'token',
  'access_token',
  'refresh_token',
  'id_token',
  'api_key',
  'client_secret',
  'private_key',
  'session',
  'assertion',
  'authorization',
  'cookie',
  'set_cookie',
  'set-cookie',
  'hmac',
  'signature',
]);

export const REDACT_VALUE = '***';

const PINO_REDACT_CONTAINERS: readonly string[] = Object.freeze([
  'req.headers',
  'request.headers',
  'res.headers',
  'response.headers',
  'req.body',
  'request.body',
  'req.query',
  'request.query',
]);

const BEARER_PATTERN = /bearer\s+[A-Za-z0-9._\-+/=]{8,}/gi;
const JWT_PATTERN = /eyJ[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}/g;
const AWS_AKIA = /AKIA[0-9A-Z]{16}/g;
const AWS_ASIA = /ASIA[0-9A-Z]{16}/g;
const GCP_KEY = /AIza[0-9A-Za-z_-]{35}/g;
const GITHUB_PAT = /gh[pousr]_[A-Za-z0-9]{36,255}/g;
const SLACK_TOKEN = /xox[abprs]-[A-Za-z0-9-]{10,}/g;
const PEM_BEGIN = '-----BEGIN ';
const PEM_END = '-----END ';
const PEM_TAIL = '-----';

export const MAX_FIELD_BYTES = (() => {
  const v = process.env.CARACAL_LOG_MAX_FIELD_BYTES;
  if (v) {
    const n = Number.parseInt(v, 10);
    if (Number.isFinite(n)) return n;
  }
  return 8192;
})();

export function isSecretKey(name: string): boolean {
  const lower = name.toLowerCase();
  return SECRET_KEYS.some((k) => lower.includes(k));
}

export function buildPinoRedactPaths(): string[] {
  const paths = new Set<string>();
  for (const k of SECRET_KEYS) {
    for (const c of PINO_REDACT_CONTAINERS) {
      paths.add(`${c}["${k}"]`);
      paths.add(`${c}["${k.toUpperCase()}"]`);
    }
  }
  return Array.from(paths);
}

/**
 * Scrub Bearer tokens, JWT-shaped substrings, and common cloud secret patterns
 * (AWS, GCP, GitHub, Slack, PEM keys) from a string. Cheap on the common path.
 */
export function redactString(s: string): string {
  if (s.length < 16) return s;
  let out = redactPemBlocks(s);
  out = out.replace(BEARER_PATTERN, `Bearer ${REDACT_VALUE}`);
  out = out.replace(JWT_PATTERN, REDACT_VALUE);
  out = out.replace(AWS_AKIA, REDACT_VALUE);
  out = out.replace(AWS_ASIA, REDACT_VALUE);
  out = out.replace(GCP_KEY, REDACT_VALUE);
  out = out.replace(GITHUB_PAT, REDACT_VALUE);
  out = out.replace(SLACK_TOKEN, REDACT_VALUE);
  return out;
}

function redactPemBlocks(s: string): string {
  let copyStart = 0;
  let pendingBegin = -1;
  let begin = s.indexOf(PEM_BEGIN);
  let end = s.indexOf(PEM_END);
  const chunks: string[] = [];

  while (begin !== -1 || end !== -1) {
    if (begin !== -1 && (end === -1 || begin < end)) {
      const label = readPemLabel(s, begin + PEM_BEGIN.length);
      if (label && isPrivateKeyPemLabel(label.value) && pendingBegin === -1) pendingBegin = begin;
      begin = s.indexOf(PEM_BEGIN, label?.end ?? begin + PEM_BEGIN.length);
      continue;
    }

    const label = readPemLabel(s, end + PEM_END.length);
    if (label && isPrivateKeyPemLabel(label.value) && pendingBegin !== -1) {
      chunks.push(s.slice(copyStart, pendingBegin), REDACT_VALUE);
      copyStart = label.end;
      pendingBegin = -1;
    }
    end = s.indexOf(PEM_END, label?.end ?? end + PEM_END.length);
  }

  if (chunks.length === 0) return s;
  chunks.push(s.slice(copyStart));
  return chunks.join('');
}

function readPemLabel(s: string, start: number): { value: string; end: number } | null {
  const end = s.indexOf(PEM_TAIL, start);
  if (end === -1) return null;
  return { value: s.slice(start, end), end: end + PEM_TAIL.length };
}

function isPrivateKeyPemLabel(label: string): boolean {
  if (!label.endsWith('PRIVATE KEY')) return false;
  for (let i = 0; i < label.length; i++) {
    const code = label.charCodeAt(i);
    if (code !== 32 && (code < 65 || code > 90)) return false;
  }
  return true;
}

export function truncateString(s: string): string {
  if (MAX_FIELD_BYTES <= 0 || s.length <= MAX_FIELD_BYTES) return s;
  return s.slice(0, MAX_FIELD_BYTES) + '…[truncated]';
}

function serializeError(err: Error): Record<string, unknown> {
  const out: Record<string, unknown> = { name: err.name, message: err.message };
  if (err.stack) out.stack = err.stack;
  const code = (err as Error & { code?: unknown }).code;
  if (code !== undefined) out.code = code;
  const cause = (err as Error & { cause?: unknown }).cause;
  if (cause !== undefined) out.cause = cause instanceof Error ? serializeError(cause) : cause;
  return out;
}

/**
 * Returns a deep copy of `value` with any field whose key matches SECRET_KEYS
 * replaced with REDACT_VALUE, string values scrubbed of token-like patterns
 * and truncated to MAX_FIELD_BYTES, and Error instances flattened to plain
 * JSON-friendly objects.
 */
export function redact<T>(value: T): T {
  if (value === null || value === undefined) return value;
  if (typeof value === 'string') return truncateString(redactString(value)) as unknown as T;
  if (value instanceof Error) return serializeError(value) as unknown as T;
  if (Array.isArray(value)) return value.map(redact) as unknown as T;
  if (typeof value !== 'object') return value;
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
    out[k] = isSecretKey(k) ? REDACT_VALUE : redact(v);
  }
  return out as T;
}

export type TraceContext = { traceId?: string; spanId?: string };

const traceStore = new AsyncLocalStorage<TraceContext>();

/** Run `fn` with the provided trace context bound to the async chain. */
export function runWithTrace<R>(tc: TraceContext, fn: () => R): R {
  return traceStore.run(tc, fn);
}

/** Enter a trace context for the remainder of the current async chain. */
export function bindTrace(tc: TraceContext): void {
  traceStore.enterWith(tc);
}

/** Returns the trace context bound to the current async chain, or undefined. */
export function getTraceContext(): TraceContext | undefined {
  return traceStore.getStore();
}

/** Parse a W3C traceparent header (`version-traceid-spanid-flags`). */
export function parseTraceparent(h: string | undefined): TraceContext {
  if (!h) return {};
  const parts = h.split('-');
  if (parts.length < 4) return {};
  if (parts[1].length !== 32 || parts[2].length !== 16) return {};
  return { traceId: parts[1], spanId: parts[2] };
}

/** Returns a pino-compatible mixin function that injects trace fields. */
export function traceMixin(): () => Record<string, unknown> {
  return () => {
    const tc = getTraceContext();
    const out: Record<string, unknown> = {};
    if (tc?.traceId) out.trace_id = tc.traceId;
    if (tc?.spanId) out.span_id = tc.spanId;
    return out;
  };
}

export type Logger = {
  level: Level;
  debug: (msg: string, fields?: Record<string, unknown>) => void;
  info: (msg: string, fields?: Record<string, unknown>) => void;
  warn: (msg: string, fields?: Record<string, unknown>) => void;
  error: (msg: string, fields?: Record<string, unknown>) => void;
  fatal: (msg: string, fields?: Record<string, unknown>) => void;
  with: (fields: Record<string, unknown>) => Logger;
  metrics: () => DevLogMetrics;
};

export type DevLogMetrics = {
  emitted: number;
  dropped: number;
  sampled: number;
};

const metricsState = { emitted: 0, dropped: 0, sampled: 0 };

export function devLogMetrics(): DevLogMetrics {
  return { ...metricsState };
}

const DEBUG_SAMPLE_N = (() => {
  const v = process.env.CARACAL_LOG_SAMPLE_DEBUG;
  if (v) {
    const n = Number.parseInt(v, 10);
    if (Number.isFinite(n) && n > 0) return n;
  }
  return 1;
})();

let debugCounter = 0;

export interface CreateLoggerOptions {
  level?: Level;
  hostname?: string;
  pid?: number;
  version?: string;
  env?: string;
  stream?: Writable;
}

let cachedHostname: string | null = null;
function host(): string {
  if (cachedHostname === null) {
    try { cachedHostname = hostname(); } catch { cachedHostname = 'unknown'; }
  }
  return cachedHostname;
}

function processBaseFields(): Record<string, unknown> {
  return {
    hostname: host(),
    pid: process.pid,
    version: process.env.CARACAL_VERSION || process.env.npm_package_version || 'dev',
    env: process.env.CARACAL_ENV || process.env.NODE_ENV || 'development',
  };
}

function envLevel(): Level {
  const raw = (process.env.CARACAL_LOG_LEVEL || process.env.LOG_LEVEL || '').toLowerCase();
  if (raw === 'debug' || raw === 'info' || raw === 'warn' || raw === 'error' || raw === 'fatal') return raw;
  return 'info';
}

export function createLogger(service: string, levelOrOpts?: Level | CreateLoggerOptions): Logger {
  const opts: CreateLoggerOptions = typeof levelOrOpts === 'string' ? { level: levelOrOpts } : (levelOrOpts ?? {});
  const level: Level = opts.level ?? envLevel();
  const baseFields: Record<string, unknown> = {
    service,
    ...processBaseFields(),
    ...(opts.hostname !== undefined ? { hostname: opts.hostname } : {}),
    ...(opts.pid !== undefined ? { pid: opts.pid } : {}),
    ...(opts.version !== undefined ? { version: opts.version } : {}),
    ...(opts.env !== undefined ? { env: opts.env } : {}),
  };
  return makeLogger(baseFields, level, opts.stream ?? process.stderr);
}

function makeLogger(
  bound: Record<string, unknown>,
  level: Level,
  stream: NodeJS.WritableStream,
): Logger {
  const threshold = LEVEL_VALUE[level];
  const emit = (msgLevel: Level, msg: string, fields?: Record<string, unknown>): void => {
    if (LEVEL_VALUE[msgLevel] < threshold) return;
    if (msgLevel === 'debug' && DEBUG_SAMPLE_N > 1) {
      debugCounter = (debugCounter + 1) >>> 0;
      if (debugCounter % DEBUG_SAMPLE_N !== 0) {
        metricsState.sampled++;
        return;
      }
    }
    const tc = getTraceContext();
    const traceFields: Record<string, unknown> = {};
    if (tc?.traceId) traceFields.trace_id = tc.traceId;
    if (tc?.spanId) traceFields.span_id = tc.spanId;
    const safe = fields ? (redact(fields) as Record<string, unknown>) : undefined;
    const line = JSON.stringify({
      level: msgLevel,
      time: new Date().toISOString(),
      ...bound,
      ...traceFields,
      msg,
      ...safe,
    }) + '\n';
    const ok = stream.write(line);
    if (ok) metricsState.emitted++;
    else metricsState.dropped++;
  };
  return {
    level,
    debug: (msg, fields) => emit('debug', msg, fields),
    info: (msg, fields) => emit('info', msg, fields),
    warn: (msg, fields) => emit('warn', msg, fields),
    error: (msg, fields) => emit('error', msg, fields),
    fatal: (msg, fields) => emit('fatal', msg, fields),
    with: (fields) => makeLogger({ ...bound, ...(redact(fields) as Record<string, unknown>) }, level, stream),
    metrics: () => devLogMetrics(),
  };
}
